### PR TITLE
feat: add plugin support to ngu-flow

### PR DESCRIPTION
**Pull request by sheikalthaf**
Introducing Plugin support, main focus is to tree shake the features and easy to use

```ts
export interface FlowPlugin {
  onInit?(data: FlowComponent): void;
  afterInit?(data: FlowComponent): void;
}
```

In this PR currently two plugins added.
1. Fit to window
2. scroll into view

Usage example

```html
<ngu-flow [config]="config">
...
</ngu-flow>
```
```ts
export class DemoComponent {
   plugins = {
      scroll: new ScrollIntoView('1'),
      fitToWindow: new FitToWindow(true),
      arrange: new Arrangements(),
   };
   config: FlowConfig = {
      Arrows: true,
      ArrowSize: 20,
      Plugins: this.plugins,
  };

  focusNode(id: string) {
    this.plugins.scroll.focus(id);
  }

  fitToWindow() {
    this.plugins.fitToWindow.fitToWindow();
  }
  
  autoArrange() {
    this.plugins.arrange.arrange();
  }
}
```

This is the basic example of the plugin usage.

### CODE REVIEW

**Comment by santoshyadavdev** [comment]
Can we add some description what we mean by plugin and which plugin we are going to support?

**Comment by sheikalthaf** [comment]
@santoshyadavdev Yes, for now I'm just exploring how we can add plugin support so that i can be more versatile. Can you throw some idea about general plugin, how they designed?

**Comment by sheikalthaf** [comment]
@santoshyadavdev can you check this proposal for plugin

File: projects/flow/src/lib/plugins/fit-to-window.spec.ts
Comment (santoshyadavdev) [inline projects/flow/src/lib/plugins/fit-to-window.spec.ts:1]: _functions generally are used for private functions, can you remove _ if possible.

File: projects/flow/src/lib/flow-child.component.ts [hunk 1]
```diff
@@ -95,9 +95,9 @@ export class FlowChildComponent implements OnInit, OnChanges, OnDestroy {
     });
 
     this.positionChange.subscribe((x) => {
-      const { left, top } = this.flow.zRect;
+      // const { left, top } = this.flow.zRect;
       // if (!this.position) console.log(this.position);
```
Comment (santoshyadavdev) [inline projects/flow/src/lib/flow-child.component.ts:99]: delete the commented code

File: projects/flow/src/lib/flow-interface.ts
Comment (santoshyadavdev) [inline projects/flow/src/lib/flow-interface.ts:1]: this is circular dependency right? we are importing this file in `FlowComponent` and 'FlowComponent` here I think its better to merge both and move this code in FlowComponent

File: projects/flow/src/lib/flow.component.ts [hunk 2]
```diff
@@ -22,9 +22,10 @@ import {
   FlowDirection,
   DotOptions,
   ArrowPathFn,
+  FlowConfig,
+  FlowPlugin,
 } from './flow-interface';
 import { blendCorners, flowPath, bezierPath, blendCorners1 } from './svg';
-import { FitToWindow } from './fit-to-window';
 
 const BASE_SCALE_AMOUNT = 0.05;
```
Comment (santoshyadavdev) [inline projects/flow/src/lib/flow.component.ts:30]: we can make it more flexible by exposing this as injectiontoken thought?

File: projects/flow/src/lib/flow.component.ts [hunk 3]
```diff
@@ -397,101 +383,27 @@ export class FlowComponent
...
-          `#${arrow.id}`
-        ) as SVGPathElement;
-        if (pathElement) {
-          pathElement.setAttribute('d', arrow.d);
-        }
-      });
-    });
-
-    this.flow.connections.updateDotVisibility(this.oldChildObj());
+    // this.flow.connections = new Connections(this.list, this.flow.direction);
```
Comment (santoshyadavdev) [inline projects/flow/src/lib/flow.component.ts:390]: remove the commented code

File: projects/flow/src/lib/plugins/arrangements.spec.ts [hunk 4]
```diff
@@ -61,7 +68,7 @@ describe('Arrangements2', () => {
       '7': { x: 660, y: 600, id: '7', deps: ['5'] },
       '8': { x: 660, y: 900, id: '8', deps: ['6', '7'] },
     };
-    const actual = Object.fromEntries(arrangements.autoArrange());
+    const actual = Object.fromEntries(arrangements._autoArrange());
```
Comment (santoshyadavdev) [inline projects/flow/src/lib/plugins/arrangements.spec.ts:71]: same her lets remove _ from functions

File: projects/flow/src/lib/plugins/arrangements.ts [hunk 5]
```diff
@@ -162,6 +194,7 @@ export class Arrangements2 {
     for (const item of this.list) {
       newItems.set(item.position.id, item.position);
     }
+    console.log([...newItems.values()]);
```
Comment (santoshyadavdev) [inline projects/flow/src/lib/plugins/arrangements.ts:197]: remove console

File: src/app/demo/demo-one.component.ts [hunk 6]
```diff
@@ -63,15 +87,43 @@ export class DemoOneComponent implements AfterViewInit {
...
 
   constructor() {
     this.list = structuredClone(FLOW_LIST);
+    this.selectedNode.valueChanges.subscribe((id) => {
+      this.plugins.scroll.focus(id);
+    });
+
+    // setTimeout(() => {
+    //   this.flowComponent.scrollIntoView('3');
+    // });
```
Comment (santoshyadavdev) [inline src/app/demo/demo-one.component.ts:112]: remove commented code

**Comment by santoshyadavdev** [comment]
Just added sonar cloud to this project as well

File: projects/flow/src/lib/plugins/fit-to-window.spec.ts
Comment (sheikalthaf) [reply projects/flow/src/lib/plugins/fit-to-window.spec.ts:1]: Yes, it is a private functions only, only for testing purpose i just made in public. Do you have any alternative to access this functions in testcases

File: projects/flow/src/lib/plugins/arrangements.spec.ts (hunk 4 above)
Comment (sheikalthaf) [reply projects/flow/src/lib/plugins/arrangements.spec.ts:44]: Just for testing purpose made it public and added _

Comment (santoshyadavdev) [reply projects/flow/src/lib/plugins/arrangements.spec.ts:44]: yes it should be testes with public methods, call the public methods and check

**Comment by sheikalthaf** [comment]
@santoshyadavdev addressed most of the comments and also solved the issues reported by sonar

**Comment by sheikalthaf** [comment]
@santoshyadavdev this PR is ready for review, can you do the final review once

**Review by santoshyadavdev — COMMENTED** [review]
we should remove the private apis, but let me know if we want to merge this and want to take care of this later.

**Comment by sheikalthaf** [comment]
@santoshyadavdev for now we are good with the plugin implementation, later I'll change the testcases and avoid public methods.

**Comment by santoshyadavdev** [comment]
Cool let's merge it

**Comment by sheikalthaf** [comment]
@santoshyadavdev  Need your approval to merge

**Review by santoshyadavdev — APPROVED** [review]

**Pull request merged**